Crawling unable to stop

I used crawljax(version 4.1) on websites with `BrowserType.REMOTE` and limited states. When the number of states is reached, the program should have stopped. However, it continues running.

Here is my code:
```java
CrawljaxConfiguration.CrawljaxConfigurationBuilder builder = CrawljaxConfiguration.builderFor(URL);
builder.crawlRules().setFormFillMode(CrawlRules.FormFillMode.RANDOM);
builder.crawlRules().clickDefaultElements();
builder.setUnlimitedRuntime();
builder.setMaximumStates(2);
builder.setUnlimitedCrawlDepth();
builder.crawlRules().clickElementsInRandomOrder(false);
builder.crawlRules().waitAfterReloadUrl(WAIT_TIME_AFTER_RELOAD, TimeUnit.MILLISECONDS);
builder.crawlRules().waitAfterEvent(WAIT_TIME_AFTER_EVENT, TimeUnit.MILLISECONDS);
builder.setBrowserConfig(BrowserConfiguration.remoteConfig(1, "http://localhost:4444/wd/hub"));
CrawljaxRunner crawljax = new CrawljaxRunner(builder.build());
crawljax.call();
System.out.println("end");
```

And the console output:
```shell
16:52:47.214 [main] WARN  com.crawljax.core.plugin.Plugins - No plugins loaded. There will be no output
16:52:47.226 [main] INFO  org.openqa.selenium.Capabilities - Using `new ChromeOptions()` is preferred to `DesiredCapabilities.chrome()`
16:52:49.045 [Forwarding newSession on session null to remote] INFO  o.o.s.remote.ProtocolHandshake - Detected dialect: OSS
16:53:00.830 [main] INFO  c.c.core.UnfiredCandidateActions - There are 1 states with unfired actions
16:53:00.833 [main] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 1
16:53:00.834 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 1 states with unfired actions
16:53:04.775 [main] INFO  com.crawljax.core.CrawlController - Received shutdown notice. Reason is Maximum states passed
16:53:04.775 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 2
16:53:04.777 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state1 added to the StateMachine.
16:53:04.777 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 1
16:53:04.783 [pool-1-thread-1] ERROR com.crawljax.core.CrawlTaskConsumer - Could not complete state crawl: org.openqa.selenium.remote.UnreachableBrowserException: Error communicating with the remote browser. It may have died.
......
	at com.crawljax.browser.WebDriverBackedEmbeddedBrowser.wrapWebDriverExceptionIfConnectionException(WebDriverBackedEmbeddedBrowser.java:917)
	at com.crawljax.browser.WebDriverBackedEmbeddedBrowser.throwIfConnectionException(WebDriverBackedEmbeddedBrowser.java:924)
	at com.crawljax.browser.WebDriverBackedEmbeddedBrowser.getStrippedDomWithoutIframeContent(WebDriverBackedEmbeddedBrowser.java:658)
	at com.crawljax.core.CandidateElementExtractor.extract(CandidateElementExtractor.java:121)
	at com.crawljax.core.Crawler.parseCurrentPageForCandidateElements(Crawler.java:555)
	at com.crawljax.core.Crawler.inspectNewDom(Crawler.java:531)
	at com.crawljax.core.Crawler.inspectNewState(Crawler.java:498)
	at com.crawljax.core.Crawler.crawlThroughActions(Crawler.java:448)
	at com.crawljax.core.Crawler.execute(Crawler.java:169)
	at com.crawljax.core.CrawlTaskConsumer.handleTask(CrawlTaskConsumer.java:78)
	at com.crawljax.core.CrawlTaskConsumer.pollAndHandleCrawlTasks(CrawlTaskConsumer.java:70)
	at com.crawljax.core.CrawlTaskConsumer.call(CrawlTaskConsumer.java:46)
	at com.crawljax.core.CrawlTaskConsumer.call(CrawlTaskConsumer.java:15)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.openqa.selenium.remote.UnreachableBrowserException: Error communicating with the remote browser. It may have died.
Build info: version: '3.141.5', revision: 'd54ebd709a', time: '2018-11-06T11:42:16'
System info: host: 'HermCdeMacBook-Pro.local', ip: 'fe80:0:0:0:14d6:93f5:6a11:4168%en0', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.15.1', java.version: '1.8.0_181'
Driver info: driver.version: RemoteWebDriver
Capabilities {acceptInsecureCerts: false, acceptSslCerts: false, applicationCacheEnabled: false, browserConnectionEnabled: false, browserName: chrome, chrome: {chromedriverVersion: 2.46.628411 (3324f4c8be9ff2..., userDataDir: /var/folders/2t/6cz2m_r15h7...}, cssSelectorsEnabled: true, databaseEnabled: false, goog:chromeOptions: {debuggerAddress: localhost:63927}, handlesAlerts: true, hasTouchScreen: false, javascriptEnabled: true, locationContextEnabled: true, mobileEmulationEnabled: false, nativeEvents: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: MAC, platformName: MAC, proxy: Proxy(), rotatable: false, setWindowRect: true, strictFileInteractability: false, takesHeapSnapshot: true, takesScreenshot: true, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unexpectedAlertBehaviour: ignore, unhandledPromptBehavior: ignore, version: 73.0.3683.75, webStorageEnabled: true, webdriver.remote.sessionid: 6f76f4e8d27c3dbe0245df53d8e...}
Session ID: 6f76f4e8d27c3dbe0245df53d8ebae0f
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:573)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:609)
	at org.openqa.selenium.remote.RemoteWebDriver.getPageSource(RemoteWebDriver.java:438)
	at com.crawljax.browser.WebDriverBackedEmbeddedBrowser.getStrippedDomWithoutIframeContent(WebDriverBackedEmbeddedBrowser.java:655)
	... 14 common frames omitted
Caused by: java.io.InterruptedIOException: thread interrupted
	at okio.Timeout.throwIfReached(Timeout.java:145)
	at okio.Okio$1.write(Okio.java:76)
	at okio.AsyncTimeout$1.write(AsyncTimeout.java:180)
	at okio.RealBufferedSink.flush(RealBufferedSink.java:224)
	at okhttp3.internal.http1.Http1Codec.finishRequest(Http1Codec.java:166)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:84)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at org.openqa.selenium.remote.internal.OkHttpClient$Factory$1.lambda$createClient$1(OkHttpClient.java:152)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at org.openqa.selenium.remote.internal.OkHttpClient.execute(OkHttpClient.java:103)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:155)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:552)
	... 17 common frames omitted
16:53:04.786 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 1 states with unfired actions
16:53:06.707 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 3
16:53:06.707 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state2 added to the StateMachine.
16:53:06.707 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 1
16:53:06.752 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 2 states with unfired actions
16:53:08.146 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 4
16:53:08.146 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state3 added to the StateMachine.
16:53:08.146 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 2
16:53:08.191 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 3 states with unfired actions
16:53:09.000 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 5
16:53:09.000 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state4 added to the StateMachine.
16:53:09.000 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 3
16:53:09.043 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 3 states with unfired actions
16:53:10.952 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 6
16:53:10.952 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state5 added to the StateMachine.
16:53:10.952 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 1
16:53:10.985 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 4 states with unfired actions
16:53:12.240 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - CLONE State detected: state6 and state2 are the same.
16:53:13.577 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 7
16:53:13.577 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state7 added to the StateMachine.
16:53:13.577 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 2
16:53:13.615 [pool-1-thread-1] INFO  c.c.core.CandidateElementExtractor - Unreadable external link http://***
16:53:13.620 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 5 states with unfired actions
16:53:15.901 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 8
16:53:15.901 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state9 added to the StateMachine.
16:53:15.902 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 3
16:53:15.933 [pool-1-thread-1] INFO  c.c.core.CandidateElementExtractor - Unreadable external link http://***
16:53:15.941 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 5 states with unfired actions
16:53:17.966 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - Crawl depth is now 1
16:53:19.115 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 9
16:53:19.115 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state10 added to the StateMachine.
16:53:19.115 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 2
16:53:19.155 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 5 states with unfired actions
16:53:19.780 [main] INFO  com.crawljax.core.CrawlController - Shutdown process complete
end
16:53:20.953 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - Crawl depth is now 1
16:53:22.116 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - Crawl depth is now 2
16:53:22.878 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 10
16:53:22.878 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state11 added to the StateMachine.
16:53:22.878 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 3
16:53:22.911 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 5 states with unfired actions
16:53:23.652 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 11
16:53:23.652 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state12 added to the StateMachine.
16:53:23.652 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 4
16:53:23.685 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 6 states with unfired actions
16:53:24.415 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 12
16:53:24.415 [pool-1-thread-1] INFO  com.crawljax.core.state.StateMachine - State state13 added to the StateMachine.
16:53:24.415 [pool-1-thread-1] INFO  com.crawljax.core.Crawler - New DOM is a new state! crawl depth is now 5
16:53:24.456 [pool-1-thread-1] INFO  c.c.core.UnfiredCandidateActions - There are 6 states with unfired actions
16:53:26.446 [pool-1-thread-1] INFO  c.c.c.state.InMemoryStateFlowGraph - Number of states in the graph is now 13
......
```

The console output means that `main` thread has finished but `pool-1-thread-1` is still working. 
<br/>
By viewing the source code, I have tried to find out where the bug lies:
When the number of states is reached, `CrawlController` uses `executor.shutdownNow()` to shutdown the thread pool. The interrupt signal will be passed to each thread. We expect that the`try-catch` block in the function `CrawlTaskConsumer::call()` will catch the exception and end the loop. However, when I use `RemoteWebDriver`, it depends on `okhttp` which will response the interrupt signal and finally throws `UnreachableBrowserException`.

<br/>
Maybe we can declare an atomic variable in the class `ExitNotifier` used to mark whether the loop can be over instead of using `InterruptedException`?

--- a/core/src/main/java/com/crawljax/browser/EmbeddedBrowser.java
+++ b/core/src/main/java/com/crawljax/browser/EmbeddedBrowser.java
@@ -5,7 +5,7 @@ import com.crawljax.core.configuration.PreCrawlConfiguration;
 import com.crawljax.core.state.Eventable;
 import com.crawljax.core.state.Identification;
 import com.crawljax.forms.FormInput;
-import org.openqa.selenium.ElementNotVisibleException;
+import org.openqa.selenium.ElementNotInteractableException;
 import org.openqa.selenium.WebDriver;
 import org.openqa.selenium.WebElement;
 
@@ -39,7 +39,7 @@ public interface EmbeddedBrowser {
 	 * @return if fails.
 	 */
 	boolean fireEventAndWait(Eventable event)
-			throws ElementNotVisibleException, InterruptedException;
+			throws ElementNotInteractableException, InterruptedException;
 
 	/**
 	 * Removes the stripped items from {@link PreCrawlConfiguration#getFilterAttributeNames()}.

--- a/core/src/main/java/com/crawljax/browser/WebDriverBackedEmbeddedBrowser.java
+++ b/core/src/main/java/com/crawljax/browser/WebDriverBackedEmbeddedBrowser.java
@@ -286,12 +286,12 @@ public final class WebDriverBackedEmbeddedBrowser implements EmbeddedBrowser {
 	 * @throws InterruptedException when interrupted during the wait.
 	 */
 	private boolean fireEventWait(WebElement webElement, Eventable eventable)
-			throws ElementNotVisibleException, InterruptedException {
+			throws ElementNotInteractableException, InterruptedException {
 		switch (eventable.getEventType()) {
 			case click:
 				try {
 					webElement.click();
-				} catch (ElementNotVisibleException e) {
+				} catch (ElementNotInteractableException e) {
 					throw e;
 				} catch (WebDriverException e) {
 					throwIfConnectionException(e);
@@ -429,7 +429,7 @@ public final class WebDriverBackedEmbeddedBrowser implements EmbeddedBrowser {
 	 */
 	@Override
 	public synchronized boolean fireEventAndWait(Eventable eventable)
-			throws ElementNotVisibleException, NoSuchElementException, InterruptedException {
+			throws ElementNotInteractableException, NoSuchElementException, InterruptedException {
 		try {
 
 			boolean handleChanged = false;
@@ -462,7 +462,7 @@ public final class WebDriverBackedEmbeddedBrowser implements EmbeddedBrowser {
 				browser.switchTo().defaultContent();
 			}
 			return result;
-		} catch (ElementNotVisibleException | NoSuchElementException e) {
+		} catch (ElementNotInteractableException | NoSuchElementException e) {
 			throw e;
 		} catch (WebDriverException e) {
 			throwIfConnectionException(e);

--- a/core/src/main/java/com/crawljax/core/Crawler.java
+++ b/core/src/main/java/com/crawljax/core/Crawler.java
@@ -17,7 +17,7 @@ import com.crawljax.util.ElementResolver;
 import com.crawljax.util.UrlUtils;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
-import org.openqa.selenium.ElementNotVisibleException;
+import org.openqa.selenium.ElementNotInteractableException;
 import org.openqa.selenium.NoSuchElementException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -358,15 +358,16 @@ public class Crawler {
 		boolean isFired = false;
 		try {
 			isFired = browser.fireEventAndWait(eventToFire);
-
+			
+		} catch (ElementNotInteractableException | NoSuchElementException e) {
 			// Handle hidden anchor tags
-			if (!isFired && crawlRules.isCrawlHiddenAnchors() && eventToFire.getElement() != null
+			if (crawlRules.isCrawlHiddenAnchors() && eventToFire.getElement() != null
 					&& "A".equals(eventToFire.getElement().getTag())) {
 				isFired = visitAnchorHrefIfPossible(eventToFire);
+			} else {
+				LOG.debug("Ignoring invisible element {}", eventToFire.getElement());
 			}
-		} catch (ElementNotVisibleException | NoSuchElementException e) {
-			LOG.debug("Ignoring invisible element {}", eventToFire.getElement());
-			isFired = false;
+			
 		} catch (InterruptedException e) {
 			LOG.debug("Interrupted during fire event");
 			interruptThread();

--- a/core/src/main/java/com/crawljax/core/Crawler.java
+++ b/core/src/main/java/com/crawljax/core/Crawler.java
@@ -363,10 +363,9 @@ public class Crawler {
 			if (!isFired && crawlRules.isCrawlHiddenAnchors() && eventToFire.getElement() != null
 					&& "A".equals(eventToFire.getElement().getTag())) {
 				isFired = visitAnchorHrefIfPossible(eventToFire);
-			} else {
-				LOG.debug("Ignoring invisible element {}", eventToFire.getElement());
 			}
 		} catch (ElementNotVisibleException | NoSuchElementException e) {
+			LOG.debug("Ignoring invisible element {}", eventToFire.getElement());
 			isFired = false;
 		} catch (InterruptedException e) {
 			LOG.debug("Interrupted during fire event");

--- a/core/src/main/java/com/crawljax/core/Crawler.java
+++ b/core/src/main/java/com/crawljax/core/Crawler.java
@@ -358,13 +358,16 @@ public class Crawler {
 		boolean isFired = false;
 		try {
 			isFired = browser.fireEventAndWait(eventToFire);
-		} catch (ElementNotVisibleException | NoSuchElementException e) {
-			if (crawlRules.isCrawlHiddenAnchors() && eventToFire.getElement() != null
+
+			// Handle hidden anchor tags
+			if (!isFired && crawlRules.isCrawlHiddenAnchors() && eventToFire.getElement() != null
 					&& "A".equals(eventToFire.getElement().getTag())) {
 				isFired = visitAnchorHrefIfPossible(eventToFire);
 			} else {
 				LOG.debug("Ignoring invisible element {}", eventToFire.getElement());
 			}
+		} catch (ElementNotVisibleException | NoSuchElementException e) {
+			isFired = false;
 		} catch (InterruptedException e) {
 			LOG.debug("Interrupted during fire event");
 			interruptThread();

--- a/core/src/test/java/com/crawljax/core/plugin/OnFireEventFailedPluginTest.java
+++ b/core/src/test/java/com/crawljax/core/plugin/OnFireEventFailedPluginTest.java
@@ -55,7 +55,7 @@ public class OnFireEventFailedPluginTest {
 	@Test
 	public void testFireEventFailedHasBeenExecuted() throws CrawljaxException {
 		controller.call();
-		assertThat("The FireEventFailed Plugin has been executed the correct amount of times",
+		assertThat("The FireEventFailed Plugin has not been executed the correct number of times",
 				hits.get(), is(2));
 	}
 

--- a/plugins/crawloverview-plugin/pom.xml
+++ b/plugins/crawloverview-plugin/pom.xml
@@ -47,7 +47,7 @@
 		<dependency>
 			<groupId>com.fasterxml.jackson.core</groupId>
 			<artifactId>jackson-databind</artifactId>
-			<version>2.9.7</version>
+			<version>[2.9.10.1,)</version>
 		</dependency>
 		<dependency>
 			<groupId>com.fasterxml.jackson.datatype</groupId>

--- a/pom.xml
+++ b/pom.xml
@@ -29,7 +29,7 @@
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
         <slf4j.version>1.8.0-beta2</slf4j.version>
         <logback.version>1.3.0-alpha4</logback.version>
-        <selenium.version>3.141.5</selenium.version>
+        <selenium.version>4.0.0-alpha-2</selenium.version>
         <jetty.version>9.4.14.v20181114</jetty.version>
         <guice.version>4.2.2</guice.version>
     </properties>
@@ -96,7 +96,7 @@
         <dependency>
             <groupId>io.github.bonigarcia</groupId>
             <artifactId>webdrivermanager</artifactId>
-            <version>3.1.1</version>
+            <version>3.7.1</version>
         </dependency>
 
         <dependency>

